# Imports

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Import files
path = r'/Users/jorisjan/Documents/CareerFoundry/Instacart Basket Analysis'
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_derived.pkl'))
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'), index_col = False)

## Create a subset of the first million rows of df_ords_prods_merg

In [4]:
df = df_ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 18)

In [6]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slower day 2        Average orders  
2  Regularly busy    Slower day 2        Average orders  
3      Least busy    Slower day 1        Average orders  
4      Least busy    Slower day 1        Average orders  
5  Regularly busy  Regularly busy        Average orders  
6  Regularly busy   Busiest day 2        Average orders  
7  Regularly busy   Busiest day 2        Average orders  
8  Regularly busy   Busiest day 2        Average orders  
9      Least busy    Slower day 1        Average orders

In [7]:
# Group 'df' dataframe by 'product_name' column
df.groupby('product_name')

In [8]:
# Calculate the mean of the “order_number” column grouped by the “department_id” column
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
# use transform() function to create a new column containing the max frequency of the “order_number” column
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/dt/sqcrp8bd1cx8p99whrf_c9wr0000gn/T/ipykernel_18652/4202709721.py:2: FutureWarning: The provided callable <function amax at 0x1131b4a40> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [10]:
df_ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda        77              7     9.0   both  Mid-range product   
11         Soda        77              7     9.0   both  Mid-range product   
12         Soda        77              7     9.0   both  Mid-range product   
13         Soda        77              7     9.0   both  Mid-range product   
14         Soda        77              7     9.0   both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy        Average orders         10  
1   Regularly busy    Slower day 2        Average orders         10  
2   Regularly busy    Slower day 2        Average orders         10  
3       Least busy    Slower day 1        Average orders         10  
4  

# 2. Find the aggregated mean of the “order_number” column grouped by “department_id” for the entire dataframe.

In [11]:
# Calculate the mean of the “order_number” column grouped by the “department_id” column
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3. How do the results for the entire dataframe differ from those of the subset?

Departments: The entire dataframe includes data for 21 departments (department_id 1 to 21), while the subset includes data for 8 departments (department_id 4, 7, 13, 14, 16, 17, 19, and 20). This exclusion means that any analysis or conclusions drawn from the subset are not representative

Mean values comparison:
Department 4: Increased from 17.811403 to 18.825780 in the subset
Department 7: Increased from 17.225802 to 17.472355 in the subset
Department 13: Increased from 16.583536 to 17.993423 in the subset
Department 14: Increased from 16.773669 to 19.246334 in the subset
Department 16: Increased from 17.665606 to 19.463012 in the subset
Department 17: Decreased from 15.694469 to 11.294069 in the subset
Department 19: Increased slightly from 17.177343 to 19.305237 in the subset
Department 20: Increased from 16.473447 to 17.599636 in the subset

The subset data generally shows higher mean values for most departments compared to the entire dataframe, with the exception of Department 17, which shows a significant decrease

# 4. Create a loyalty flag for existing customers using the transform() and loc() functions

In [12]:
# create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value

In [13]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [14]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [15]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
# Run the value_counts() function to check the values in the 'loyalty_flag' column.
df_ords_prods_merge['loyalty_flag'].value_counts()

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [17]:
# check that everything was interpreted correctly and that the right flags were assigned
df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(15)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22

# 5. Determining the difference between the spending habits of the three types of customers identified

In [18]:
# Calculate the mean of the “prices” column grouped by the “loyalty_flag” column
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

# 6. Create a spending flag for each user based on the average price across all their orders

In [19]:
# use transform() function to create a new column containing the mean amount of the “prices” column
df_ords_prods_merge['mean_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/dt/sqcrp8bd1cx8p99whrf_c9wr0000gn/T/ipykernel_18652/3237836770.py:2: FutureWarning: The provided callable <function mean at 0x1131b5760> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['mean_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [20]:
# create a flag that assigns a “spending” label to a user ID based on its corresponding mean spending value

In [21]:
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [22]:
df_ords_prods_merge.loc[df_ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'

In [23]:
# Run the value_counts() function to check the values in the 'loyalty_flag' column.
df_ords_prods_merge['spending_flag'].value_counts()

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [24]:
# check that everything was interpreted correctly and that the right flags were assigned
df_ords_prods_merge[['user_id', 'spending_flag', 'order_number', 'mean_price']].head(25)

user_id spending_flag  order_number  mean_price
0         1   Low spender             1    6.367797
1         1   Low spender             2    6.367797
2         1   Low spender             3    6.367797
3         1   Low spender             4    6.367797
4         1   Low spender             5    6.367797
5         1   Low spender             6    6.367797
6         1   Low spender             7    6.367797
7         1   Low spender             8    6.367797
8         1   Low spender             9    6.367797
9         1   Low spender            10    6.367797
10       15   Low spender            15    3.980556
11       15   Low spender            17    3.980556
12       15   Low spender            18    3.980556
13       15   Low spender            21    3.980556
14       15   Low spender            22    3.980556
15       19   Low spender             2    8.372059
16       19   Low spender             5    8.372059
17       19   Low spender             7    8.372059
18       21   Low spender            10    7.012683
19       31   Low spender            10    8.137793
20       31   Low spender            17    8.137793
21       43   Low spender             6    7.444898
22       43   Low spender             9    7.444898
23       52   Low spender             1    6.356287
24       52   Low spender             2    6.356287

# 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column

In [25]:
# use transform() function to create a new column containing the median of 'days_since_prior_order' column
df_ords_prods_merge['median_days_spo'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/dt/sqcrp8bd1cx8p99whrf_c9wr0000gn/T/ipykernel_18652/876547740.py:2: FutureWarning: The provided callable <function median at 0x11332b100> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['median_days_spo'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [26]:
# create a flag that assigns a “regular” label to a user ID based on its corresponding median period since prior order

In [27]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_spo'] > 20, 'regular_flag'] = 'Non-frequent customer'

In [28]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_spo'] <= 20) & (df_ords_prods_merge['median_days_spo'] > 10), 'regular_flag'] = 'Regular customer'

In [29]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_spo'] <= 10, 'regular_flag'] = 'Frequent customer'

In [30]:
# Run the value_counts() function to check the values in the 'regular_flag' column.
df_ords_prods_merge['regular_flag'].value_counts()

regular_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
nan                             5
Name: count, dtype: int64

In [31]:
# check that everything was interpreted correctly and that the right flags were assigned
df_ords_prods_merge[['user_id', 'regular_flag', 'order_number', 'median_days_spo']].head(15)

user_id           regular_flag  order_number  median_days_spo
0         1  Non-frequent customer             1             20.5
1         1  Non-frequent customer             2             20.5
2         1  Non-frequent customer             3             20.5
3         1  Non-frequent customer             4             20.5
4         1  Non-frequent customer             5             20.5
5         1  Non-frequent customer             6             20.5
6         1  Non-frequent customer             7             20.5
7         1  Non-frequent customer             8             20.5
8         1  Non-frequent customer             9             20.5
9         1  Non-frequent customer            10             20.5
10       15      Frequent customer            15             10.0
11       15      Frequent customer            17             10.0
12       15      Frequent customer            18             10.0
13       15      Frequent customer            21             10.0
14       15      Frequent customer            22             10.0

# 9. Export df_ords_prods_merge as a pickle file and store it correctly in your “Prepared Data” folder.

In [32]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_grouped&agg.pkl'))

# Turn price values >100 into NaNs

In [33]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [34]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_last.pkl'))